adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,280 @@
+This situation introduces several issues:
+
+* It is unclear which classes are intended as public APIs versus internal implementations.
+* There is no clear boundary between what an extension consumer can depend on and what they should not use.
+* Some extensions mix SPI and implementation code, reducing modularity and increasing the chance of unwanted coupling.
+* The need for build item re-use or coordination between extensions leads to ad-hoc patterns.
+* The word _deployment_ is there for historic reasons, but does not make a lot of sense. `build` would be better.
+
+More recently, we introduced dev-only modules, further increasing complexity.
+
+To address these issues and prepare for long-term maintainability (and possibly Java Platform Module System (JPMS) adoption), we propose a more structured layout for extension modules.
+
+== Decision
+
+Extensions may adopt a standardized structure, composed of one mandatory and several optional modules, depending on their complexity and integration needs:
+
+* `runtime`: Contains the runtime implementation.
+It may expose public APIs if needed but should prefer delegation to the `runtime-api` module (See <<the-place-of-the-public-api>>).
+This module also contains the `recorders` and other runtime logic.
+* `runtime-api`: Contains public APIs intended to be consumed by other extensions or application code.
+This module should have minimal dependencies.
+Depending on an `runtime-api` module does not transitively include the full extension (i.e., it avoids pulling in runtime or build logic). See <<coupling-api-and-spi>> for more details.
+* `build`: Contains build-time logic, including `BuildSteps`.
+This module may define internal build items.
+* `build-spi`: Contains build-time APIs intended to be reused across multiple extensions for integration purposes.
+Extensions contributing to or using build items from other extensions should depend on `build-spi` modules.
+* `runtime-dev`: Contains runtime classes used exclusively in dev mode (e.g., for the Dev UI). This avoids shipping dev-only classes into production artifacts.
+* `codestart`: Contains the _Codestart_ templates for the extension. See https://quarkus.io/guides/extension-codestart[Codestart Guide] for more details.
+* `cli`: Contains CLI commands related to the extension. This is optional and should be used only if the extension provides specific CLI functionality.
+* `codegen`: Contains code generation utilities such as code pre-processors invoked at build time.
+
+Only the `runtime` and `build` modules are mandatory; the others are optional based on the extension's needs.
+
+
+.New structure for Quarkus extensions
+[source,tree]
+----
+my-extension/
+├── runtime/
+├── runtime-api/
+├── build/
+├── build-spi/
+├── runtime-dev/
+├── codegen/
+├── cli/
+└── codestart/
+----
+
+NOTE: `deployment` and `deployment-spi` have been renamed to `build` and `build-spi`. This is a big change, but Maven relocation will mitigate the breakage.
```

Comment:
I think we need to take a deep breath here and think about all the consequences of this change. It has relatively low value and could require a massive amount of work.

If we actually need relocations (which I'm not sure of), our BOM will be even more gigantic and it will slow down our Maven builds (it's already in the way as it is).

Also, we need to think about how we will support extensions that have not yet been migrated to the new model. This change impacts us but also impacts the whole ecosystem.

It's THE change that will trigger work on EVERY extension out there.